### PR TITLE
Add docs, unit tests, and refactor repro test infra

### PR DESCRIPTION
**Pull request by charleskawczynski**
This PR adds docs, unit tests, and refactors the reproducibility test infrastructure. In adding unit tests and documentation, I've realized that some of the terminology around the repro tests needs revised for certain terminology to be well defined.

In particular, if we consider the situation where commits are reverted, we can end up with multiple and diverging paths (multiple paths with the same reference counter, but made at different periods in development). For example, if our reference counter history looks like (top being most recent):

```julia
137
     /PR10 # is PR6 even valid to compare against, here?
136
     /PR9
     /PR10 # should this compare with PRs 4 and 5?
135  # Reverted
     /PR8 # should this compare with commits in PRs 1,2,3?
138
     /PR7
137
     /PR6
136
     /PR4
     /PR5
135
     /PR1
     /PR2
     /PR3
```
We could clean up older commits, but I think a safer option is to change the `latest_comparable_paths` to use the (newly) added function `compute_bins`. I originally used the term `bundle` (e.g., `compute_bundles`), which I'm almost leaning towards returning to. It's a longer name but sounds more...tidy.

This was previously not an issue when `latest_comparable_paths` was `latest_comparable_path`, because we always grabbed the most _recent_ matched reference to compare against, eliminating this potential issue.

I'll update the README once we land on a design that is coherent, but I think it might take a bit more refactoring to get there.

### CODE REVIEW

File: reproducibility_tests/reproducibility_utils.jl [hunk 1]
```diff
@@ -118,57 +159,156 @@ function latest_comparable_paths(;
...
-    f = joinpath(path, "ref_counter.jl")
-    if !isfile(f)
-        return "ref_counter.jl does not exist"
-    else
-        ref_counter = parse(Int, first(readlines(f)))
-        return "ref_counter: $ref_counter"
+"""
+    invalid_reference_folders(; root_path)
+
+Returns all subfolders in `root_path`
```
Comment (Sbozzolo) [inline reproducibility_tests/reproducibility_utils.jl:165]:
```suggestion
Return all subfolders in `root_path`
```

File: reproducibility_tests/reproducibility_utils.jl [hunk 2]
```diff
@@ -118,57 +159,156 @@ function latest_comparable_paths(;
...
-        for i in 1:N
-            path = sorted_paths[i]
-            ref_counter_file = joinpath(path, "ref_counter.jl")
-            if !isfile(ref_counter_file)
-                push!(paths_to_delete, path)
+"""
+    compute_bins(root_path::String)
+    compute_bins(sorted_paths::Vector{String})
+
+Returns a vector of reproducibility bins.
```
Comment (Sbozzolo) [inline reproducibility_tests/reproducibility_utils.jl:182]:
```suggestion
Return a vector of reproducibility bins.
```

File: reproducibility_tests/reproducibility_utils.jl [hunk 3]
```diff
@@ -118,57 +159,156 @@ function latest_comparable_paths(;
...
-        for i in 1:length(paths_to_delete)
-            f = paths_to_delete[i]
-            @info "     (File, date): ($(f), $(Dates.unix2datetime(stat(f).mtime))). Reason: $(reason(f))"
-        end
-        for i in 1:length(paths_to_delete)
-            rm(paths_to_delete[i]; recursive = true, force = true)
+        push!(bins, paths_per_bin)
+    end
+    return bins
+end
```
Comment (Sbozzolo) [inline reproducibility_tests/reproducibility_utils.jl:232]: I think the logic of this function hard to follow. Some of the the issues are: 
- there are two nested and interacting `while true`  (both of which start with an exit condition, meaning that we can just turn them into `while condition`)
- variable names are obscure
- the intent of the various instructions is not very clear to me

I tried writing it differently, but I am not sure if I am preserving the logic:
```suggestion
function compute_bins(sorted_paths::Vector{String})
    num_paths = length(sorted_paths)
    bins = Vector{String}[]
    path_index = 1
    while path_index <= num_paths
            paths_per_bin = String[]
            while path_index <= num_paths
                first_path = sorted_paths[path_index]
                first_ref_counter = read_ref_counter(joinpath(first_path, "ref_counter.jl"))
                push!(paths_per_bin, first_path)
                path_index += 1
                current_path = sorted_paths[path_index]
                current_ref_counter = read_ref_counter(joinpath(p, "ref_counter.jl"))
                if current_ref_counter == first_ref_counter
                    push!(paths_per_bin, current_path)
                    path_index += 1
                else
                    break
                end
        end
        push!(bins, paths_per_bin)
    end
    return bins
end
```
This version also doesn't make the intent very clear

File: reproducibility_tests/reproducibility_utils.jl [hunk 4]
```diff
@@ -118,57 +159,156 @@ function latest_comparable_paths(;
...
+
+"""
+    get_reference_paths_to_delete(;
+        root_path,
+        keep_n_comparable_states = 5,
+        keep_n_bins_back = 7,
+    )
+
+Returns a list of folders to delete.
```
Comment (Sbozzolo) [inline reproducibility_tests/reproducibility_utils.jl:241]:
```suggestion
Return a list of folders to delete.
```

File: reproducibility_tests/reproducibility_utils.jl [hunk 5]
```diff
@@ -118,57 +159,156 @@ function latest_comparable_paths(;
...
+
+Our reference folders are saved, and can
+therefore build up significantly.
```
Comment (Sbozzolo) [inline reproducibility_tests/reproducibility_utils.jl:244]:
```suggestion
Our reference folders are saved on a shared folder on the cluster, and can
therefore build up significantly and take a lot of storage.
```

**Review by Sbozzolo — APPROVED** [review]
Thanks! I think this is a big improvement in transparency on what the reproducibility tests are doing.

I just left one comment for a function whole logic I have hard time following

File: reproducibility_tests/reproducibility_utils.jl (hunk 5 above)
Comment (charleskawczynski) [reply reproducibility_tests/reproducibility_utils.jl:244]: So, one reason I left this generic is for re-useability / portability. I'd like to remove, or make as many things not actually depend on central (e.g., the unit tests can now exercise this locally, which doesn't have anything to do with central).

File: reproducibility_tests/reproducibility_utils.jl (hunk 3 above)
Comment (charleskawczynski) [reply reproducibility_tests/reproducibility_utils.jl:232]: Hm, I'm getting BoundsError with this version, also, needing to change `current_ref_counter = read_ref_counter(joinpath(p, "ref_counter.jl"))` to `current_ref_counter = read_ref_counter(joinpath(current_path, "ref_counter.jl"))`

Comment (charleskawczynski) [reply reproducibility_tests/reproducibility_utils.jl:232]: I can try to improve the local variable names

Comment (charleskawczynski) [reply reproducibility_tests/reproducibility_utils.jl]: I've updated the names and changed the `while` conditions

File: reproducibility_tests/reproducibility_utils.jl (hunk 5 above)
Comment (charleskawczynski) [reply reproducibility_tests/reproducibility_utils.jl]: I kept the `and take a lot of storage` piece.

Comment (Sbozzolo) [reply reproducibility_tests/reproducibility_utils.jl]: Okay, I just wanted to tell readers why we care about removing things.

Comment (charleskawczynski) [reply reproducibility_tests/reproducibility_utils.jl]: I think we can do that where it's called in `move_output.jl`
